service/round.go

```diff
@@ -121,65 +96,44 @@ func (r *round) submit(key, challenge []byte) error {
 	return r.challengesDb.Put(key, challenge, &opt.WriteOptions{Sync: true})
 }
 
-func (r *round) numChallenges() int {
-	iter := r.challengesDb.NewIterator(nil, nil)
-	defer iter.Release()
-
-	var num int
-	for iter.Next() {
-		num++
-	}
-
-	return num
-}
-
-func (r *round) isEmpty() bool {
-	iter := r.challengesDb.NewIterator(nil, nil)
-	defer iter.Release()
-	return !iter.Next()
-}
-
 func (r *round) execute(ctx context.Context, end time.Time, minMemoryLayer uint) error {
 	logger := logging.FromContext(ctx).WithFields(log.String("round", r.ID))
 	logger.Info("executing until %v...", end)
 
-	r.executionStarted = time.Now()
+	r.ExecutionStarted = time.Now()
 	if err := r.saveState(); err != nil {
 		return err
 	}
 
-	close(r.executionStartedChan)
-
-	var err error
-	r.execution.Members, r.execution.Statement, err = r.calcMembersAndStatement()
-	if err != nil {
+	if members, statement, err := r.calcMembersAndStatement(); err != nil {
 		return err
+	} else {
+		r.Execution.Members, r.Execution.Statement = members, statement
 	}
 
 	if err := r.saveState(); err != nil {
 		return err
 	}
 
-	r.execution.NumLeaves, r.execution.NIP, err = prover.GenerateProof(
+	numLeaves, nip, err := prover.GenerateProof(
 		ctx,
 		r.datadir,
-		hash.GenLabelHashFunc(r.execution.Statement),
-		hash.GenMerkleHashFunc(r.execution.Statement),
+		hash.GenLabelHashFunc(r.Execution.Statement),
+		hash.GenMerkleHashFunc(r.Execution.Statement),
 		end,
-		r.execution.SecurityParam,
+		r.Execution.SecurityParam,
```

Comment:
Same here, `Execution` can be modified from outside of `round`.

Reply:
It could be modified outside of `round` either way as Go has no notion of "private vs public", just "unexported vs exported on a package level". `round struct` is unexported so it's only available inside the `service` pkg.